cache-filter/src/filter/http.rs

```diff
@@ -148,37 +150,72 @@ impl CacheFilter {
 
         let current_time = self.get_current_time().duration_since(UNIX_EPOCH)?;
+        let mut updated_cache = false;
+        let mut rate_limited = false;
+        let max_tries = self.config.max_tries;
 
         add_hierarchy_to_metrics(&app.metric_hierarchy, &mut self.req_data.metrics);
-        match limit_check_and_update_application(&self.req_data, app, &current_time) {
-            Ok(()) => {
-                // request is not rate-limited and application is updated
-                if !set_application_to_cache(&self.cache_key.as_string(), &app, true, None) {
-                    self.update_cache_from_singleton = true;
+
+        // In case of CAS mismatch, new application needs to be fetched and modified again.
+        for num_try in 1..max_tries {
+            match limit_check_and_update_application(&self.req_data, app, app_cas, &current_time) {
+                Ok(()) => {
+                    // App is not rate-limited and updated in cache.
+                    info!(
+                        "ctxt {}: request is allowed to pass the filter",
+                        self.context_id
+                    );
+                    if !self.report_to_singleton(queue_id, &current_time) {
+                        // TODO: Handle MQ failure here
+                        // Update local cache
+                        // Report to 3scale and get new state using authrep endpoint
+                    }
+                    updated_cache = true;
+                    self.resume_http_request();
                 }
-                info!(
-                    "ctxt {}: request is allowed to pass the filter",
-                    self.context_id
-                );
-                if !self.report_to_singleton(queue_id, &current_time) {
-                    // TODO: Handle MQ failure here
-                    // Update local cache
-                    // Report to 3scale and get new state using authrep endpoint
+                Err(UpdateMetricsError::RateLimited) => {
+                    info!("ctxt {}: request is rate-limited", self.context_id);
+                    self.send_http_response(429, vec![], Some(b"Request rate-limited.\n"));
+                    // no need to retry if already rate-limted
+                    rate_limited = true;
+                    break;
                 }
-                self.resume_http_request();
-            }
-            Err(UpdateMetricsError::RateLimited) => {
-                info!("ctxt {}: request is rate-limited", self.context_id);
-                // TODO: add how many similar requests can be accepted.
-                self.send_http_response(429, vec![], Some(b"Request rate-limited.\n"));
+                Err(UpdateMetricsError::CacheUpdateFail) => {
+                    info!(
+                        "ctxt {}: try #{} failed to set application to cache",
+                        self.context_id, num_try
+                    );
+                    if num_try < max_tries {
```

Comment:
Actually, change is required in the start `for num_try in 1..(max_tries+1)`. We need to check that because we don't want to get a new application if that's the last try for the set.

Reply:
ah yes. for loop should go up to `max_retries+1` and if is required.

Reply:
What would happen if a user sets `max_tries` to the value `4294967296`?